tamagotchi1/src/lib.rs

```diff
@@ -1,81 +1,207 @@
+            0,
+        ).expect("Failed to share TmgEvent");
+    }
+
+    fn play(&mut self) {
+        let tamagotchi = unsafe {
+            STATE.as_mut().expect("The contract is not initialized")
+        };
+
+        assert_eq!(msg::source(), tamagotchi.owner, "Only owner can feed the tamagotchi");
+        assert!(self.happy < 5000, "Tamagotchi not enough hungry");
+
+        let lost = (block_height() - tamagotchi.entertained_block) * Self::ENERGY_PER_BLOCK;
+
+        self.happy += Self::FILL_PER_ENTERTAINMENT;
+
+        self.happy = match lost {
+            hl if self.happy < hl => 0,
+            hl if MAX_NUM < hl => MAX_NUM,
+            _ => self.happy - lost,
+        };
+
+        debug!("play: {}", self.happy);
+
+        tamagotchi.entertained = self.happy;
+        tamagotchi.entertained_block = block_height();
+
+        msg::reply(
+            TmgEvent::Entertained,
+            0,
+        ).expect("Failed to share TmgEvent");
+    }
+
+    fn sleep(&mut self) {
+        let tamagotchi = unsafe {
+            STATE.as_mut().expect("The contract is not initialized")
+        };
+
+        assert_eq!(msg::source(), tamagotchi.owner, "Only owner can feed the tamagotchi");
+        assert!(self.rested < 5000, "Tamagotchi not enough hungry");
+
+        // calculating and normalizing hunger level
+        let lost = (block_height() - tamagotchi.rested_block) * Self::BOREDOM_PER_BLOCK;
+
+
+        self.rested += Self::FILL_PER_SLEEP;
+        self.rested = match lost {
+            hl if self.rested < hl => 0,
+            hl if MAX_NUM < hl => MAX_NUM,
+            _ => self.rested - lost,
+        };
+
+        debug!("sleep: {}", self.rested);
+
+        tamagotchi.rested = self.rested;
+        tamagotchi.rested_block = block_height();
+
+        msg::reply(
+            TmgEvent::Slept,
+            0,
+        ).expect("Failed to share TmgEvent");
+    }
+}
 
 #[no_mangle]
 extern "C" fn init() {
-    let greeting: String = msg::load()
-        .expect("Can't decode an init message");
-    debug!("Program was initialized with message {:?}",
-        greeting);
-    unsafe { GREETING = Some(greeting) };
+    let owner = msg::source();
+
+    let name: String = String::from_utf8(
+        msg::load_bytes().expect("Can't load tamagotchi name")
+    ).expect("Can't decode tamagotchi name");
+
+    // let name: String = msg::load().expect("Can't load tamagotchi name");
+
+    let date_of_birth = block_timestamp();
+
+    let bh = block_height();
+
+
+    unsafe {
+        MOOD = Some(Mood {
+            fed: INT_NUM,
+            happy: INT_NUM,
+            rested: INT_NUM,
+        });
+    };
+
+    unsafe {
+        STATE = Some(Tamagotchi {
+            owner,
+            name,
+            date_of_birth,
+            fed: INT_NUM,
+            rested: INT_NUM,
+            entertained: INT_NUM,
+            fed_block: bh.clone(),
+            entertained_block: bh.clone(),
+            rested_block: bh.clone(),
```

Comment:
No need to clone `Copy` type.
```suggestion
            fed_block: bh,
            entertained_block: bh,
            rested_block: bh,
```